{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # unzip dataset\n",
    "# !pwd\n",
    "# !ls -lh ../0-data/0-dataset\n",
    "# !ls -lh ../0-data/1-prepare_data\n",
    "# # mouse 10090\n",
    "# !gunzip -c ../0-data/0-dataset/10090.protein.links.v11.0.txt.gz > ../0-data/1-prepare_data/10090-ppi.txt\n",
    "# !gunzip -c ../0-data/0-dataset/uniprot-filtered-organism__Mus+musculus+\\(Mouse\\)+\\[10090\\]_+AND+revie--.xlsx.gz > ../0-data/1-prepare_data/10090-uniprot.xlsx\n",
    "# # human 9606\n",
    "# !gunzip -c ../0-data/0-dataset/9606.protein.links.v11.0.txt.gz > ../0-data/1-prepare_data/9606-ppi.txt\n",
    "# !gunzip -c ../0-data/0-dataset/uniprot-filtered-organism__Homo+sapiens+\\(Human\\)+\\[9606\\]_+AND+review--.xlsx.gz > ../0-data/1-prepare_data/9606-uniprot.xlsx\n",
    "\n",
    "# !ls -lh ../0-data/1-prepare_data\n",
    "\n",
    "# !mkdir ./temp  \n",
    "# %cd ./temp\n",
    "# !gunzip -c ../../0-data/0-dataset/uniprot-filtered-organism__Homo+sapiens+\\(Human\\)+\\[9606\\]_+AND+review--.fasta.gz > ./9606.fa\n",
    "# !gunzip -c ../../0-data/0-dataset/uniprot-filtered-organism__Mus+musculus+\\(Mouse\\)+\\[10090\\]_+AND+revie--.fasta.gz > ./10090.fa\n",
    "# !diamond makedb --in ./9606.fa  -d 9606 && rm ./9606.fa\n",
    "# !diamond makedb --in ./10090.fa -d 10090 && rm ./10090.fa\n",
    "# !gunzip -c ../../0-data/0-dataset/uniprot-filtered-organism__Homo+sapiens+\\(Human\\)+\\[9606\\]_+AND+review--query.fasta.gz > ./9606-query.fa\n",
    "# !gunzip -c ../../0-data/0-dataset/uniprot-filtered-organism__Mus+musculus+\\(Mouse\\)+\\[10090\\]_+AND+revie--qurey.fasta.gz > ./10090-query.fa\n",
    "\n",
    "# # multiple sequence alignment\n",
    "# !diamond blastp -q ./9606-query.fa -d 9606.dmnd  -o ../../0-data/1-prepare_data/9606-diamod.tsv --very-sensitive -b8 -c1\n",
    "# !diamond blastp -q ./10090-query.fa -d 10090.dmnd  -o ../../0-data/1-prepare_data/10090-diamod.tsv --very-sensitive -b8 -c1\n",
    "\n",
    "# !ls -lh ../../0-data/1-prepare_data/\n",
    "# %cd ../\n",
    "# !rm -r ./temp/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" \n",
    "命名规范\n",
    "    模块与包：module_name, package_name\n",
    "    类名与异常：ClassName, ExceptionName\n",
    "    方法名：method_name, function_name\n",
    "    变量：instance_var_name, function_parameter_name, local_var_name, GLOBAL_VAR_NAME\n",
    "\n",
    "作用域：\n",
    "    protected: _protected_var, _protected_function, _ProtectedClass\n",
    "    private: __\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pickle\n",
    "import collections\n",
    "\n",
    "from tqdm import tqdm\n",
    "from pandas.core.frame import DataFrame, Series\n",
    "from typing import Dict, Tuple, Sequence\n",
    "\n",
    "# output control\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "pd.set_option(\"max_columns\", 1000)\n",
    "pd.set_option(\"max_row\", 300)\n",
    "pd.set_option(\"display.float_format\", lambda x: '%.5f' % x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE_DIR = \"../0-data/1-prepare_data/\"\n",
    "OUT_DIR = \"../0-data/2-build_graphs/\"\n",
    "OBO_NAME = \"../0-data/0-dataset/go.obo\"\n",
    "RESERVE_PPI = False\n",
    "ALPHA_FACTOR = 1500\n",
    "\n",
    "INPUT_DATASET = {\n",
    "    \"human\": {\n",
    "        \"id\": \"9606\",\n",
    "        \"ppi\": \"9606-ppi.txt\",\n",
    "        \"uniprot\": \"9606-uniprot.xlsx\",\n",
    "        \"diamond\": \"9606-diamond.tsv\"\n",
    "    },\n",
    "    \"mouse\": {\n",
    "        \"id\": \"10090\",\n",
    "        \"ppi\": \"10090-ppi.txt\",\n",
    "        \"uniprot\": \"10090-uniprot.xlsx\",\n",
    "        \"diamond\": \"10090-diamond.tsv\"\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# handle single type of input\n",
    "def handle_uniprot(df_uniprot: DataFrame, reserve_ppi: bool=False) -> DataFrame:\n",
    "    df_uniprot.columns = ['uniprot_id', 'string_id', 'sequence', 'labels']\n",
    "    # 清洗 string_id 字段\n",
    "    df_uniprot['string_id'] = df_uniprot['string_id'].map(lambda x: x.strip(';') if isinstance(x, str) else x)\n",
    "    \n",
    "    # 由于部分对比方法无法处理无PPI信息的数据，因此去除空字段\n",
    "    # 注意本方法支持无PPI信息的数据，注释掉此句可以保留这些节点\n",
    "    if not reserve_ppi:\n",
    "        df_uniprot = df_uniprot.drop(df_uniprot[df_uniprot['string_id'].map(lambda x: False if isinstance(x, str) else True)].index).reset_index(drop=True)\n",
    "    \n",
    "    return df_uniprot\n",
    "\n",
    "\n",
    "def handle_ppi(df_ppi: DataFrame, string2uniprot: dict, reserve_ppi: bool=False) -> DataFrame:\n",
    "    # 映射函数\n",
    "    get_uniprot = lambda x: string2uniprot.get(x, float('nan'))\n",
    "    # 保留所有PPI数据时使用此函数，对于存在映射的，优先使用uniprotid，否则使用stringid\n",
    "    get_uniprot_reserve_x = lambda x: string2uniprot.get(x, x)\n",
    "    get_fun = get_uniprot_reserve_x if reserve_ppi else get_uniprot\n",
    "    \n",
    "    # 清洗 ppi 数据\n",
    "    df_ppi['protein1'] = df_ppi['protein1'].map(lambda x: get_fun(x))\n",
    "    df_ppi['protein2'] = df_ppi['protein2'].map(lambda x: get_fun(x))\n",
    "    \n",
    "    # df_ppi = df_ppi.drop(df_ppi[df_ppi['protein1'].map(lambda x: False if isinstance(x, str) else True)].index).reset_index(drop=True)\n",
    "    # df_ppi = df_ppi.drop(df_ppi[df_ppi['protein2'].map(lambda x: False if isinstance(x, str) else True)].index).reset_index(drop=True)\n",
    "    df_ppi = df_ppi.dropna().reset_index(drop=True)\n",
    "    \n",
    "    return df_ppi\n",
    "\n",
    "\n",
    "def handle_diamond(df_diamond: DataFrame, uniprot_ids: set, reserve_ppi: bool=False) -> DataFrame:\n",
    "    df_diamond = df_diamond[['protein1', 'protein2', 'bit_score']].copy(deep=True)\n",
    "    # 字段映射\n",
    "    df_diamond['protein1'] = df_diamond['protein1'].map(lambda x: x.split(\"|\")[1].split(\"-\")[0])\n",
    "    df_diamond['protein2'] = df_diamond['protein2'].map(lambda x: x.split(\"|\")[1].split(\"-\")[0])\n",
    "    \n",
    "    # 字段清洗\n",
    "    df_diamond['protein1'] = df_diamond['protein1'].map(lambda x: x if x in uniprot_ids else float('nan'))\n",
    "    df_diamond['protein2'] = df_diamond['protein2'].map(lambda x: x if x in uniprot_ids else float('nan'))\n",
    "    \n",
    "    df_diamond = df_diamond.dropna().reset_index(drop=True)\n",
    "    return df_diamond"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocessing_data(species: str=\"human\") -> Tuple[DataFrame, DataFrame, DataFrame]:\n",
    "    \"\"\"Read source file into memory \n",
    "    \"\"\"\n",
    "    uniprot = BASE_DIR + INPUT_DATASET[species]['uniprot']\n",
    "    ppi = BASE_DIR + INPUT_DATASET[species]['ppi']\n",
    "    diamond = BASE_DIR + INPUT_DATASET[species]['diamond']\n",
    "\n",
    "    df_uniprot = pd.read_excel(uniprot)[['Entry', 'Cross-reference (STRING)', 'Sequence', 'Gene ontology IDs']]\n",
    "    df_ppi = pd.read_csv(ppi, sep=' ')\n",
    "    df_diamond = pd.read_csv(diamond, sep='\\t', header=None, names=['protein1', 'protein2', 'sequence identity percentage', 'length', 'mismatches', 'gap openings', 'start1', 'end1', 'start2', 'end2', 'e-value', 'bit_score'])\n",
    "\n",
    "    df_uniprot = handle_uniprot(df_uniprot, RESERVE_PPI)\n",
    "\n",
    "    # 映射字典\n",
    "    string2uniprot = dict(zip(df_uniprot['string_id'], df_uniprot['uniprot_id']))\n",
    "    df_ppi = handle_ppi(df_ppi, string2uniprot, RESERVE_PPI)\n",
    "\n",
    "    # 保留 id\n",
    "    uniprot_ids  = set(df_uniprot['uniprot_id'])\n",
    "    df_diamond = handle_diamond(df_diamond, uniprot_ids, RESERVE_PPI)\n",
    "    return df_uniprot, df_ppi, df_diamond"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_similar_matrix(df_ppi: DataFrame, uniprot2id: dict, n: int, out_prefix: str) -> str:\n",
    "    \n",
    "    cnt = 0\n",
    "    path = \"./\" + out_prefix + \"-ppi-mat.pkl\"\n",
    "    \n",
    "    if os.path.exists(path):\n",
    "        return path\n",
    "    similar_matrix = [[0.]*n for i in range(n)]\n",
    "\n",
    "    for _, row in tqdm(df_ppi.iterrows(), total = len(df_ppi)):\n",
    "        cnt += 1\n",
    "        if (cnt % 1000000 == 0):\n",
    "            print(row)\n",
    "        x = uniprot2id[row['protein1']]\n",
    "        y = uniprot2id[row['protein2']]\n",
    "        if (similar_matrix[x][y] != 0):\n",
    "            print(x, \" \", y)\n",
    "        similar_matrix[x][y] += row['combined_score']\n",
    "\n",
    "    cnt\n",
    "    with open(path, \"wb+\") as f:\n",
    "        pickle.dump(similar_matrix, f)\n",
    "        \n",
    "    return path\n",
    "\n",
    "def build_diamond(df_uniprot: DataFrame, df_diamond: DataFrame, uniprot2id: dict, path: str) -> list:\n",
    "    with open(path, \"rb\") as f:\n",
    "        similar_matrix = pickle.load(f)\n",
    "\n",
    "    for protein1 in tqdm(df_uniprot['uniprot_id']):\n",
    "    #     print(protein1)\n",
    "        df = df_diamond[df_diamond['protein1']==protein1]\n",
    "        x = uniprot2id[protein1]\n",
    "\n",
    "        for score in df[df['protein2'] == protein1]['bit_score']:\n",
    "            similar_matrix[x][x] = max(similar_matrix[x][x], score)\n",
    "\n",
    "        for _, row in df.iterrows():\n",
    "            y = uniprot2id[row['protein2']]\n",
    "            if x != y:\n",
    "                similar_matrix[x][y] = row['bit_score'] / similar_matrix[x][x] * ALPHA_FACTOR\n",
    "\n",
    "        similar_matrix[x][x] = 0\n",
    "\n",
    "#     with open(\"./9606-similar-mat.pkl\", \"wb+\") as f:\n",
    "#         pickle.dump(similar_matrix, f)\n",
    "    return similar_matrix\n",
    "\n",
    "def build_ppsn(df_uniprot: DataFrame, df_ppi: DataFrame, df_diamond: DataFrame, species: str=\"human\") -> DataFrame:\n",
    "    \n",
    "    out_prefix = INPUT_DATASET[species]['id']\n",
    "    df_uniprot = df_uniprot.reset_index().rename(columns={'index': 'id'})\n",
    "    uniprot2id = dict(zip(df_uniprot['uniprot_id'], df_uniprot['id']))\n",
    "    path = build_similar_matrix(df_ppi, uniprot2id, len(df_uniprot), out_prefix)\n",
    "    similar_matrix = build_diamond(df_uniprot, df_diamond, uniprot2id, path)\n",
    "    \n",
    "    protein1 = []\n",
    "    protein2 = []\n",
    "    score = []\n",
    "    n = len(df_uniprot)\n",
    "\n",
    "    for i in tqdm(range(n)):\n",
    "        for j in range(n):\n",
    "            if similar_matrix[i][j] != 0:\n",
    "                if i == j:\n",
    "                    print(\"error\")\n",
    "                    break;\n",
    "                protein1.append(i)\n",
    "                protein2.append(j)\n",
    "                score.append(similar_matrix[i][j])\n",
    "                \n",
    "    # protein-protein similarity networks\n",
    "    ppsn = DataFrame({\n",
    "        'protein1': protein1,\n",
    "        'protein2': protein2,\n",
    "        'score': score\n",
    "    })\n",
    "    \n",
    "    # output result\n",
    "    ppsn.to_csv(OUT_DIR + out_prefix + \"-ppsn-min.csv\", index=False, sep='\\t')\n",
    "    df_uniprot.to_csv(OUT_DIR + out_prefix +\"-uniprot-min.csv\", index=False, sep='\\t')\n",
    "    return ppsn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def resolve_terms():\n",
    "    \n",
    "    go_path = OBO_NAME\n",
    "    out_path = OUT_DIR + \"terms.pkl\"\n",
    "    gos = []\n",
    "    global namespace\n",
    "    namespace = collections.defaultdict(str)\n",
    "    is_a = collections.defaultdict(list)\n",
    "    part_of = collections.defaultdict(list)\n",
    "\n",
    "    # 根据规则来提取go term ，并依据其之间的依赖关系构建图谱\n",
    "    with open(go_path,'r')as f:\n",
    "        for line in f:\n",
    "            if '[Typedef]' in line:\n",
    "                break\n",
    "            if line[:5]=='id: G':                       # 构建 gos\n",
    "                line=line.strip().split()\n",
    "                gos.append(line[1])\n",
    "            elif line[:4]=='is_a':                      # 构建 is_a 关系\n",
    "                line=line.strip().split()\n",
    "                is_a[gos[-1]].append(line[1])\n",
    "            elif line[:4]=='rela' and 'part' in line:   # 构建 partof 关系\n",
    "                line=line.strip().split()\n",
    "                part_of[gos[-1]].append(line[2])\n",
    "            elif line[:5]=='names':                     # 统计子本体\n",
    "                line=line.strip().split()\n",
    "                namespace[gos[-1]]=line[1]\n",
    "\n",
    "    son_of = {\n",
    "        \"GO:0008150\": None,\n",
    "        \"GO:0005575\": None,\n",
    "        \"GO:0003674\": None\n",
    "    }\n",
    "\n",
    "    son_of = {**son_of, **is_a}\n",
    "\n",
    "    for i in part_of:\n",
    "        son_of[i].extend(part_of[i])\n",
    "\n",
    "    cross_ontology = 0\n",
    "    for k in son_of:\n",
    "        tag = False\n",
    "        if son_of[k] is None:\n",
    "            continue\n",
    "        for j in son_of[k]:\n",
    "            if namespace[k]!=namespace[j]:\n",
    "    #             print(\"error: child->{}, father->{}\".format(k, j))\n",
    "                tag = True\n",
    "        if tag:\n",
    "            cross_ontology += 1\n",
    "            \n",
    "    with open(out_path, \"wb+\") as f:\n",
    "        pickle.dump([son_of, namespace], f)\n",
    "    \n",
    "    print(\"Cross-ontology connection in prat_of: \", cross_ontology)\n",
    "    print(\"Total terms: \", len(gos))         # 47210\n",
    "    print(\"is_a relation: \", len(is_a))        # 44082\n",
    "    print(\"part_of relation: \", len(part_of))     # 8295\n",
    "    print(len(namespace))   # 47210\n",
    "    print(\"output terms: \", len(son_of))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cross-ontology connection in prat_of:  1165\n",
      "Total terms:  47210\n",
      "is_a relation:  44082\n",
      "part_of relation:  8295\n",
      "47210\n",
      "output terms:  44085\n"
     ]
    }
   ],
   "source": [
    "resolve_terms()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "human\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhangxs/.local/anaconda3/envs/dgl/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: Your version of xlrd is 1.2.0. In xlrd >= 2.0, only the xls format is supported. As a result, the openpyxl engine will be used if it is installed and the engine argument is not specified. Install openpyxl instead.\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************************** human *********************************** : \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(18560, 4)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(11098152, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(252027, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>string_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q66K14</td>\n",
       "      <td>9606.ENSP00000349291</td>\n",
       "      <td>MWLSPEEVLVANALWVTERANPFFVLQRRRGHGRGGGLTGLLVGTL...</td>\n",
       "      <td>GO:0005096; GO:0005509; GO:0006886; GO:0016021...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q9UMR3</td>\n",
       "      <td>9606.ENSP00000386170</td>\n",
       "      <td>MEFTASPKPQLSSRANAFSIAALMSSGGSKEKEATENTIKPLEQFV...</td>\n",
       "      <td>GO:0000122; GO:0000785; GO:0000977; GO:0000978...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9P031</td>\n",
       "      <td>9606.ENSP00000256151</td>\n",
       "      <td>MAPVRRSAKWRPGGIEARGEGVSTVGYRNKNVRQKTWRPNHPQAFV...</td>\n",
       "      <td>GO:0003723; GO:0005654; GO:0044267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q6PEY2</td>\n",
       "      <td>9606.ENSP00000318197</td>\n",
       "      <td>MRECISIHVGQAGVQIGNACWELYCLEHGIQPDGQMPSDKTIGGGD...</td>\n",
       "      <td>GO:0000226; GO:0000278; GO:0003924; GO:0005200...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9P016</td>\n",
       "      <td>9606.ENSP00000341657</td>\n",
       "      <td>MSRPRKRLAGTSGSDKGLSGKRTKTENSGEALAKVEDSNPQKTSAT...</td>\n",
       "      <td>GO:0005634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id             string_id  \\\n",
       "0     Q66K14  9606.ENSP00000349291   \n",
       "1     Q9UMR3  9606.ENSP00000386170   \n",
       "2     Q9P031  9606.ENSP00000256151   \n",
       "3     Q6PEY2  9606.ENSP00000318197   \n",
       "4     Q9P016  9606.ENSP00000341657   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MWLSPEEVLVANALWVTERANPFFVLQRRRGHGRGGGLTGLLVGTL...   \n",
       "1  MEFTASPKPQLSSRANAFSIAALMSSGGSKEKEATENTIKPLEQFV...   \n",
       "2  MAPVRRSAKWRPGGIEARGEGVSTVGYRNKNVRQKTWRPNHPQAFV...   \n",
       "3  MRECISIHVGQAGVQIGNACWELYCLEHGIQPDGQMPSDKTIGGGD...   \n",
       "4  MSRPRKRLAGTSGSDKGLSGKRTKTENSGEALAKVEDSNPQKTSAT...   \n",
       "\n",
       "                                              labels  \n",
       "0  GO:0005096; GO:0005509; GO:0006886; GO:0016021...  \n",
       "1  GO:0000122; GO:0000785; GO:0000977; GO:0000978...  \n",
       "2                 GO:0003723; GO:0005654; GO:0044267  \n",
       "3  GO:0000226; GO:0000278; GO:0003924; GO:0005200...  \n",
       "4                                         GO:0005634  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P84085</td>\n",
       "      <td>O43307</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P84085</td>\n",
       "      <td>O75460</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P84085</td>\n",
       "      <td>P42771</td>\n",
       "      <td>606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P84085</td>\n",
       "      <td>P07237</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P84085</td>\n",
       "      <td>O60499</td>\n",
       "      <td>267</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein1 protein2  combined_score\n",
       "0   P84085   O43307             198\n",
       "1   P84085   O75460             159\n",
       "2   P84085   P42771             606\n",
       "3   P84085   P07237             167\n",
       "4   P84085   O60499             267"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>bit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q66K14</td>\n",
       "      <td>Q66K14</td>\n",
       "      <td>2435.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q66K14</td>\n",
       "      <td>Q66K14</td>\n",
       "      <td>2390.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q66K14</td>\n",
       "      <td>Q6ZT07</td>\n",
       "      <td>1496.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q66K14</td>\n",
       "      <td>O95759</td>\n",
       "      <td>1071.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q66K14</td>\n",
       "      <td>Q0IIM8</td>\n",
       "      <td>1045.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein1 protein2  bit_score\n",
       "0   Q66K14   Q66K14 2435.00000\n",
       "1   Q66K14   Q66K14 2390.00000\n",
       "2   Q66K14   Q6ZT07 1496.00000\n",
       "3   Q66K14   O95759 1071.00000\n",
       "4   Q66K14   Q0IIM8 1045.00000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 18560/18560 [05:26<00:00, 56.81it/s]\n",
      "100%|██████████| 18560/18560 [00:41<00:00, 452.40it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mouse\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhangxs/.local/anaconda3/envs/dgl/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: Your version of xlrd is 1.2.0. In xlrd >= 2.0, only the xls format is supported. As a result, the openpyxl engine will be used if it is installed and the engine argument is not specified. Install openpyxl instead.\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*********************************** mouse *********************************** : \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(16420, 4)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(9730128, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(191350, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>string_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8VCZ3</td>\n",
       "      <td>10090.ENSMUSP00000097701</td>\n",
       "      <td>MPRHCSAAGCCTRDTRETRNRGISFHRLPKKDNPRRGLWLANCQRL...</td>\n",
       "      <td>GO:0000122; GO:0001226; GO:0003677; GO:0005634...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q62264</td>\n",
       "      <td>10090.ENSMUSP00000042988</td>\n",
       "      <td>MQVLTKRYPKNCLLTVMDRYSAVVRNMEQVVMIPSLLRDVQLSGPG...</td>\n",
       "      <td>GO:0005654; GO:0005829; GO:0006629; GO:0009617...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q6QNU9</td>\n",
       "      <td>10090.ENSMUSP00000074381</td>\n",
       "      <td>MGRYWLLPGLLLSLPLVTGWSTSNCLVTEGSRLPLVSRYFTFCRHS...</td>\n",
       "      <td>GO:0002224; GO:0004888; GO:0005887; GO:0006954...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q8BHE4</td>\n",
       "      <td>10090.ENSMUSP00000140027</td>\n",
       "      <td>MKRSLQALYCQLLSFLLTLALTKALVLAVHEPSPRESLQTLPSGSP...</td>\n",
       "      <td>GO:0005769; GO:0006898; GO:0008090; GO:0010008...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9JLF7</td>\n",
       "      <td>10090.ENSMUSP00000106625</td>\n",
       "      <td>MACQLDLLIGVIFMASPVLVISPCSSDGRIAFFRGCNLTQIPWILN...</td>\n",
       "      <td>GO:0002224; GO:0002755; GO:0004888; GO:0005149...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id                 string_id  \\\n",
       "0     Q8VCZ3  10090.ENSMUSP00000097701   \n",
       "1     Q62264  10090.ENSMUSP00000042988   \n",
       "2     Q6QNU9  10090.ENSMUSP00000074381   \n",
       "3     Q8BHE4  10090.ENSMUSP00000140027   \n",
       "4     Q9JLF7  10090.ENSMUSP00000106625   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  MPRHCSAAGCCTRDTRETRNRGISFHRLPKKDNPRRGLWLANCQRL...   \n",
       "1  MQVLTKRYPKNCLLTVMDRYSAVVRNMEQVVMIPSLLRDVQLSGPG...   \n",
       "2  MGRYWLLPGLLLSLPLVTGWSTSNCLVTEGSRLPLVSRYFTFCRHS...   \n",
       "3  MKRSLQALYCQLLSFLLTLALTKALVLAVHEPSPRESLQTLPSGSP...   \n",
       "4  MACQLDLLIGVIFMASPVLVISPCSSDGRIAFFRGCNLTQIPWILN...   \n",
       "\n",
       "                                              labels  \n",
       "0  GO:0000122; GO:0001226; GO:0003677; GO:0005634...  \n",
       "1  GO:0005654; GO:0005829; GO:0006629; GO:0009617...  \n",
       "2  GO:0002224; GO:0004888; GO:0005887; GO:0006954...  \n",
       "3  GO:0005769; GO:0006898; GO:0008090; GO:0010008...  \n",
       "4  GO:0002224; GO:0002755; GO:0004888; GO:0005149...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9DC51</td>\n",
       "      <td>Q9WUN2</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q9DC51</td>\n",
       "      <td>P11031</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9DC51</td>\n",
       "      <td>Q9JKV2</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q9DC51</td>\n",
       "      <td>P14685</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q9DC51</td>\n",
       "      <td>Q8BHC1</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein1 protein2  combined_score\n",
       "0   Q9DC51   Q9WUN2             183\n",
       "1   Q9DC51   P11031             155\n",
       "2   Q9DC51   Q9JKV2             258\n",
       "3   Q9DC51   P14685             165\n",
       "4   Q9DC51   Q8BHC1             192"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>bit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8VCZ3</td>\n",
       "      <td>Q8VCZ3</td>\n",
       "      <td>622.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q8VCZ3</td>\n",
       "      <td>Q9D305</td>\n",
       "      <td>62.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q62264</td>\n",
       "      <td>Q62264</td>\n",
       "      <td>294.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q62264</td>\n",
       "      <td>Q9CQ20</td>\n",
       "      <td>71.60000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q6QNU9</td>\n",
       "      <td>Q6QNU9</td>\n",
       "      <td>1714.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein1 protein2  bit_score\n",
       "0   Q8VCZ3   Q8VCZ3  622.00000\n",
       "1   Q8VCZ3   Q9D305   62.00000\n",
       "2   Q62264   Q62264  294.00000\n",
       "3   Q62264   Q9CQ20   71.60000\n",
       "4   Q6QNU9   Q6QNU9 1714.00000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16420/16420 [03:57<00:00, 69.08it/s]\n",
      "100%|██████████| 16420/16420 [00:32<00:00, 503.97it/s]\n"
     ]
    }
   ],
   "source": [
    "# TODO: build ppsn which retains all information of ppi \n",
    "for species in INPUT_DATASET.keys():\n",
    "    print(species)\n",
    "    df_uniprot, df_ppi, df_diamond = preprocessing_data(species)\n",
    "\n",
    "    print(\"*\"*35, species, \"*\"*35, \": \")\n",
    "    df_uniprot.shape\n",
    "    df_ppi.shape\n",
    "    df_diamond.shape\n",
    "\n",
    "    df_uniprot.head()\n",
    "    df_ppi.head()\n",
    "    df_diamond.head()\n",
    "\n",
    "    ppsn = build_ppsn(df_uniprot, df_ppi, df_diamond, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['human', 'mouse'])"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "INPUT_DATASET.keys()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
